From 3a77467c459a45e23cc9f1e751cab74cbf2be01d Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:22 -0500
Subject: [drm] i915: Kick fbdev before vgacon

Message-id: <1418852380-13061-72-git-send-email-rclark@redhat.com>
Patchwork-id: 102208
O-Subject: [RHEL7 drm 71/89] drm/i915: Kick fbdev before vgacon
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since drm-intel-fixes-2014-11-19

commit 0485c9dc24ec0939b42ca5104c0373297506b555

Author:     Daniel Vetter <daniel.vetter@ffwll.ch>
AuthorDate: Fri Nov 14 10:09:49 2014 +0100
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Mon Dec 15 14:18:50 2014 -0500

    drm/i915: Kick fbdev before vgacon

    It's magic, but it seems to work.

    This fixes a regression introduced in

    commit 1bb9e632a0aeee1121e652ee4dc80e5e6f14bcd2
    Author: Daniel Vetter <daniel.vetter@ffwll.ch>
    Date:   Tue Jul 8 10:02:43 2014 +0200

        drm/i915: Only unbind vgacon, not other console drivers

    My best guess is that the vga fbdev driver falls over if we rip out
    parts of vgacon. Hooray.

    Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=82439
    Cc: stable@vger.kernel.org (v3.16+)
    Reported-and-tested-by: Lv Zheng <lv.zheng@intel.com>
    Signed-off-by: Daniel Vetter <daniel.vetter@intel.com>
    Acked-by: Chris Wilson <chris@chris-wilson.co.uk>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/i915_dma.c b/drivers/gpu/drm/i915/i915_dma.c
index 644a7a9..bb61cff 100644
--- a/drivers/gpu/drm/i915/i915_dma.c
+++ b/drivers/gpu/drm/i915/i915_dma.c
@@ -1662,15 +1662,17 @@ int i915_driver_load(struct drm_device *dev, unsigned long flags)
   goto out_regs;
 
  if (drm_core_check_feature(dev, DRIVER_MODESET)) {
-  ret = i915_kick_out_vgacon(dev_priv);
+  /* WARNING: Apparently we must kick fbdev drivers before vgacon,
+   * otherwise the vga fbdev driver falls over. */
+  ret = i915_kick_out_firmware_fb(dev_priv);
   if (ret) {
-   DRM_ERROR("failed to remove conflicting VGA console\n");
+   DRM_ERROR("failed to remove conflicting framebuffer drivers\n");
    goto out_gtt;
   }
 
-  ret = i915_kick_out_firmware_fb(dev_priv);
+  ret = i915_kick_out_vgacon(dev_priv);
   if (ret) {
-   DRM_ERROR("failed to remove conflicting framebuffer drivers\n");
+   DRM_ERROR("failed to remove conflicting VGA console\n");
    goto out_gtt;
   }
  }
-- 
1.7.1